Tidy RangeSet

bachase opened the pull request:
@wufeipeng fix an issue in #2109 in the unused `getPrev` function of `RangeSet`.  This is an alternate fix, which eliminates several unused members of `RangeSet` and increases test coverage of the remaining member functions.

mellery451 commented on src/test/basics/RangeSet_test.cpp:
```diff
@@ -26,69 +26,132 @@ namespace ripple {
 public:
-    RangeSet createPredefinedSet ()
+    void
+    testSetAndHas()
```
these are good tests - the only other thing I'd like to see is `setRange` with values that overlap and the front/back/middle of an existing range. If that's already there and I missed it, then 👍

bachase replied:
Good suggestion.  I will add.

bachase replied:
Thanks @mellery451, your suggestion uncovered a bug.  I'm actually now more tempted to look at [Boost ICL](http://www.boost.org/doc/libs/1_64_0/libs/icl/doc/html/index.html) after all to avoid these type of errors when rolling our own.

bachase commented:
Sorry to do this to you @ximinez and @mellery451, but I decided to switch to boost after that last bug was uncovered.  I choose to have`RangeSet` class use an `interval_set` internally to keep the `LedgerMaster` code the same.

ximinez commented:
I don't mind a bit! I'll check it out ASAP.

bachase commented:
Squash and passed, thanks!

nbougalis commented on src/test/basics/RangeSet_test.cpp:
```diff
@@ -26,69 +26,175 @@ namespace ripple {
+        for (int i = 1; i <= 10; ++i)
+        {
+            if(!BEAST_EXPECT(r.hasValue(i)))
+                std::cout << i << "\n";
```
`std::cout` in a unit test?

bachase replied:
D'oh. Forgot to tidy that.

nbougalis commented on src/test/basics/RangeSet_test.cpp:
```diff
@@ -26,69 +26,175 @@ namespace ripple {
+                ((i % 10) > 6) ? (i - 1) : oneBelowRange;
 
-            BEAST_EXPECT(set.prevMissing (i) == expectedPrevMissing);
+            BEAST_EXPECT(set.prevMissing(i) == expectedPrevMissing);
```
So, `prevMissing` returns a `std::uint32_t` (which may be initialized to `static_int<std::uint32_t>(-1)`) and `expectedPrevMissing` is an `int`. This smells.

nbougalis commented on src/ripple/basics/RangeSet.h:
```diff
@@ -20,72 +20,44 @@
 
     // Add an item to the set
-    void setValue (std::uint32_t);
+    void setValue(std::uint32_t);
```
I know you went out of your way to not rename things, but since you're here, I think that some renaming makes sense: I'd suggest that `hasValue` should be `contains`, `setValue` and `setRange` should really be `add` and `clearValue` should be `remove`.

miguelportilla commented:
@bachase The new Shard code uses `lebesgue_sum`, so we will need that. A couple of deficiencies I found with this class- No way to clear it and no way to determine if it is empty other than comparing the first item to `absent` which is ugly. I highly suggest adding `clear` and `empty` functions.

bachase commented:
Thanks for the feedback.  I'm all for changing the interface as you suggest @nbougalis and @miguelportilla rather than the incremental changes I proposed. I'll also restore `lesbegue_sum`.

bachase commented:
Get out your 🍅.  `RangeSet` was ending up as a mostly trivial wrapper over `boost::icl::interval_set`, so I decided to just make it a type alias and code the non-trivial `to_string` and `prevMissing` methods as free functions.  I'll admit the `boost::icl` has a bit of a learning curve, so I'm open to wrapping more of the interface if that is preferable. 

@miguelportilla checkout https://github.com/bachase/rippled/commits/bc_shard to see the changes I made in your shard work.

miguelportilla commented:
@bachase I like the simplification of an alias and some free functions. The shard changes look good. Using `empty` and `clear` vs the previous hacks are definitely an improvement. The "tidy" became an overhaul! Thanks and nice job! 👍

bachase commented:
Rebased on 0.70.0-b6 and squashed with just the latest approach since the incremental change isn't particularly useful to see.  Note that `rs.lesbegue_sum` is replaced by boost::icl::length(rs)`.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -365,15 +365,16 @@ LedgerMaster::getFullValidatedRange (std::uint32_t& minVal, std::uint32_t& maxVa
     if (!maxVal)
         return false;
 
+    boost::optional<std::uint32_t> maybeMin;
```
I'm going with the idea that it's not worth changing the interface here to make `minVal` an `optional`.

nbougalis replied:
If anything, I think changing this to _return_ `boost::optional<std::pair<std::uint32_t,std::uint32_t>>` makes better sense.

ximinez replied:
Hm. That's a good idea. There's no reason the parameters need to be passed in to start with. Same could be said for `getValidatedRange`.

bachase replied:
I'm going to punt on that for this PR.  I think its worth considering, but better as a refactor/simplification of `LedgerMaster`.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -1478,12 +1465,9 @@ LedgerMaster::getPropertySource ()
-    }
+    ScopedLockType sl(mCompleteLock);
+    if (seq > 0)
+        mCompleteLedgers.erase(range(0u, seq - 1));
```
Nice simplification. Can `mCompleteLock` be changed into a regular `mutex` now?

bachase replied:
I *think* so, but I think the design of `LedgerMaster` doesn't enforce it well, so I was wary to change it.

ximinez commented on src/ripple/app/ledger/impl/LedgerMaster.cpp:
```diff
@@ -344,14 +344,14 @@ bool
 {
     ScopedLockType sl (mCompleteLock);
-    return mCompleteLedgers.hasValue (seq);
+    return boost::icl::contains(mCompleteLedgers, seq);
```
Is it worth adding another free function to "hide" the `boost::icl` in this one case? Probably not.

seelabs commented:
Merged into 0.80.0-b1
